### DISCUSSION TITLE
Crazyflie not running with custom controller - developing and debugging a new controller on crazyflie firmware

### DISCUSSION THREAD

**Discussion by ArthurHDN** [bitcraze/discussions#889] [General]
Hello. 
We recently acquired a swarm of crazyflies for academic research and I'm trying to implement a new controller. Its being a good experience so far and the docs and community is helps a lot, but I'm having some questions and I'm stuck for the last days so I was wondering if anyone would be able to help me. I will break down what I'm doing and the questions/problems that I'm having. 

**My goal:**
I developed a controller and I want to use a singe crazyflie to test and validate it. Since it is a controller my understanding is that I should modify the crazyflie firwmare. Just the crazyflie firmware and not the crazyradio firmware is enough if I'm getting it right. For the final implementation, I'm not sure if I want to pass velocity setpoints or setpoints of any kind to avoid my controller being implemented in a cascade strategy. Maybe just a few parameters is enough to define the path. 

Since it is my first time playing with firmwares and with the crazyflie, I'm trying to be the most baby stepped as possible.

**What I'm trying**
1) Controller as an oot app layer
After building and testing the app_hello_world, I'm putting my controller as an out ot tree controller, as per these docs:
https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/development/oot/
https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/userguides/app_layer/

So I have a `controller_outoftree.c`, a `Kbuild` file with "obj-y += controller_outoftree.o", and a `app-config` with "CONFIG_CONTROLLER_OOT=y". I can build and flash as per the docs.

For the app_hello_world, testing it was easy with cfclient. But for my controller, how can I be sure that the crazyflie is using it and not one of the others in the firmware?

From my understading, the oot controller is the controller 5. Then, using cflib (more details below), I'm setting stabilizer.controller param to 5 and I'm assuming it is working. But I can set this parameter to whatever integer I like and don't get error. Is there a way to make sure that the running controller is the one I'm adding?

2) Running the controller
The other question is how can I properly run my controller? As I said, I'm going in steps. So I didn't fully implemented my controller. I only implemented the attitude control part and I attached the crazyflie to a 3D-printed gimbal ring that allows it to rotate. Hence, I can test/tune only the attitude part. 

The controller_outoftree.c is as follows so far:
```c
#include "log.h"
#include "param.h"
#include "num.h"
#include "math3d.h"
#include "physicalConstants.h"
#include "stabilizer_types.h"

...
...


#define UPDATE_RATE RATE_100_HZ

void controllerMyController(controllerMyController_t* self,
                                control_t *control,
                                 const setpoint_t *setpoint,
                                 const sensorData_t *sensors,
                                 const state_t *state,
                                 const stabilizerStep_t stabilizerStep) {

  static float control_omega[3];
  static struct vec control_torque;
  static float control_thrust;

  // define this here, since we do body-rate control at 1000Hz below the following if statement
  float omega[3] = {0};
  omega[0] = radians(sensors->gyro.x);
  omega[1] = radians(sensors->gyro.y);
  omega[2] = radians(sensors->gyro.z);

  if (!RATE_DO_EXECUTE(UPDATE_RATE, stabilizerStep)) {
    return;
  }

  float dt = (float)(1.0f/UPDATE_RATE);

...
...
// compute the torques as a struct vec named tau

  control_thrust = 4; //anything just to rotate the properllers

  control->thrustSi = control_thrust * CF_MASS; // force to provide control_thrust
  control->torqueX = tau.x;
  control->torqueY = tau.y;
  control->torqueZ = tau.z;

  control->controlMode = controlModeForceTorque;
}


void controllerOutOfTreeInit(void) {
  controllerMyControllerInit(&g_self);
}


void controllerOutOfTree(control_t *control, const setpoint_t *setpoint, const sensorData_t *sensors, const state_t *state, const stabilizerStep_t stabilizerStep) {
  controllerMyController(&g_self, control, setpoint, sensors, state, stabilizerStep);
}


bool controllerOutOfTreeTest(void) {
  return controllerMyControllerTest(&g_self);
}
```

And this is how I'm trying to run it, with a python script that sends a dummy setpoint so it will trigger my controller. My controller doesn't look at any setpoint since the target attitude is hardcodded there, at least for now.
```python
import logging
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.utils import uri_helper

from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncLogger import SyncLogger
from cflib.positioning.motion_commander import MotionCommander

# URI to the Crazyflie to connect to
uri = uri_helper.uri_from_env(default='radio://0/100/2M/E7E7E7E704')

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)


if __name__ == '__main__':
    # Initialize the low-level drivers
    cflib.crtp.init_drivers()

    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:

        print("Yeah, I'm connected! :D")

        # # Test properllers
        # print('Testing properlers...')
        # scf.cf.param.set_value('health.startPropTest', 0)
        # scf.cf.param.set_value('health.startPropTest', 1)
        # time.sleep(8)
        # scf.cf.param.set_value('health.startPropTest', 0)
        # print('Done testing properlers')

       # Set controller to 5, estimator to 2
        scf.cf.param.set_value('stabilizer.estimator', 2)
        scf.cf.param.set_value('stabilizer.controller', 5)
        # simple_log_async(scf, lg_stab)

        print('Flying...')
        # with MotionCommander(scf) as mc:
        #     for i in range(10):
        #         print(i)
        #         mc.start_linear_motion(0.0,0.0,0.2)
        #         time.sleep(1)
        #     mc.stop()
        scf.cf.commander.send_hover_setpoint(0, 0, 0, 0)
        for i in range(50):
            print(i)
            time.sleep(0.2)
            scf.cf.commander.send_hover_setpoint(0, 0, 0, 0)
        scf.cf.commander.send_stop_setpoint()
        print('Done')

        # simple_log_async(scf, lg_stab)

        time.sleep(1)
        print("Now I will disconnect :'(")

```

However I'm getting a mixed behavior. Sometimes the crazyflie runs as crazy (no pun intented), maybe my controller isn't tuned yet, but the problem is in the sometimes.

Most of the times, it just don't do nothing. At first I had the impression that I had to reboot it, but I'm not sure. I don't see any errors when running the script. I also tried running the others controllers and sending other setpoints, but still most of the times it does nothing. Is there a way to check if the packages are being sent and consumed or if the controller is actually doing something? Any ideas on why it is moving only sometimes with my approach?

Thanks in advance.
Arthur

**Comment by krichardsson** [comment]
Hi,
I think it sounds like you are on the right track!

> From my understading, the oot controller is the controller 5. Then, using cflib (more details below), I'm setting stabilizer.controller param to 5 and I'm assuming it is working. But I can set this parameter to whatever integer I like and don't get error. Is there a way to make sure that the running controller is the one I'm adding?

Yes, that is right. You can find the definitions in [controller.c](https://github.com/bitcraze/crazyflie-firmware/blob/e176751cf1ecd6d85e37abe52338aff46968596a/src/modules/src/controller/controller.c#L25-L34) and as you can see the oot controller will be nr 5. 
The actual switching of controllers can be found in [stabilizer.c](https://github.com/bitcraze/crazyflie-firmware/blob/e176751cf1ecd6d85e37abe52338aff46968596a/src/modules/src/stabilizer.c#L229-L232). If you want to you could add a print to show when the controller changes. 

> Most of the times, it just don't do nothing. 

This might be that the Crazyflie has detected a crash and locks up to prevent bad things to happen. If you connect the client to the Crazyflie and look in the console log, take off and flip the Crazyflie up-side-down you will see "SUP: Locked, reboot required", this is the supervisor letting you know that it is locked. 
If this is the case you can turn off the tumble detection by defining the SUPERVISOR_TUMBLE_CHECK_ENABLE to false and recompile the firmware.

**Reply by ArthurHDN** [reply]
Hello.
I added the suggested change:
```
diff --git a/src/platform/interface/platform_defaults.h b/src/platform/interface/platform_defaults.h
index af8d5d0f..117b811c 100644
--- a/src/platform/interface/platform_defaults.h
+++ b/src/platform/interface/platform_defaults.h
@@ -111,7 +111,7 @@
 
 // Tumble detection enabled by default
 #ifndef SUPERVISOR_TUMBLE_CHECK_ENABLE
-    #define SUPERVISOR_TUMBLE_CHECK_ENABLE true
+    #define SUPERVISOR_TUMBLE_CHECK_ENABLE false
 #endif
```
I can confirm that setting SUPERVISOR_TUMBLE_CHECK_ENABLE to false resolved my problem and my controller can fully run with the python script above! It also clarified my question about running oot controller with nr 5. I still have to reboot the crazyflie to run the code for the second time though.

Thanks for the console hint, I also added a console callback (ref: https://www.bitcraze.io/documentation/repository/crazyflie-lib-python/master/api/cflib/crazyflie/console/). This is how it looks like:
```python
import logging
import time

import cflib.crtp
from cflib.crazyflie import Crazyflie
from cflib.crazyflie.syncCrazyflie import SyncCrazyflie
from cflib.utils import uri_helper

from cflib.crazyflie.log import LogConfig
from cflib.crazyflie.syncLogger import SyncLogger

# URI to the Crazyflie to connect to
uri = uri_helper.uri_from_env(default='radio://0/100/2M/E7E7E7E704')

# Only output errors from the logging framework
logging.basicConfig(level=logging.ERROR)

def log_console(text):
    print(f'CONSOLE: {text}')

if __name__ == '__main__':
    # Initialize the low-level drivers
    cflib.crtp.init_drivers()

    lg_stab = LogConfig(name='Stabilizer', period_in_ms=10)
    lg_stab.add_variable('stateEstimate.qw', 'float')
    lg_stab.add_variable('stateEstimate.qx', 'float')
    lg_stab.add_variable('stateEstimate.qy', 'float')
    lg_stab.add_variable('stateEstimate.qz', 'float')

    with SyncCrazyflie(uri, cf=Crazyflie(rw_cache='./cache')) as scf:
        print("Yeah, I'm connected! :D")

        # Log console
        scf.cf.console.receivedChar.add_callback(log_console)

        # # Test properllers
        # print('Testing properlers...')
        # scf.cf.param.set_value('health.startPropTest', 0)
        # scf.cf.param.set_value('health.startPropTest', 1)
        # time.sleep(8)
        # scf.cf.param.set_value('health.startPropTest', 0)
        # print('Done testing properlers')

        scf.cf.param.set_value('stabilizer.estimator', 3)
        scf.cf.param.set_value('stabilizer.controller', 5)

        scf.cf.log.add_config(lg_stab)
        # lg_stab.start()

        print('Flying...')
        try:
            for i in range(200):
                time.sleep(0.2)
                scf.cf.commander.send_hover_setpoint(0, 0, 0, 0)
        finally:
            pass
            # scf.cf.commander.send_stop_setpoint()
        print('Done')

        # lg_stab.stop()

        time.sleep(1)
        print("Now I will disconnect :'(")
```
After the controller is executed for the first time, I see this message:
```bash
Flying...
CONSOLE: CONTROLLER: Using OutOfTree (5
CONSOLE: ) controller

Done
CONSOLE: SUP: Can not fly
```
And when running for the second time without rebooting:
```
Flying...
Done
```
Am I doing something that is changing the supervisor state machine to a state other than supervisorStateReadyToFly ? Is there a way to set this state before running the controller?

Anyway, having it to run as it is atm already helped a lot. Thanks again Kristoffer!
Arthur

**Reply by krichardsson** [reply]
You need to call `cf.commander.send_notify_setpoint_stop()`
For more information see https://github.com/bitcraze/crazyflie-firmware/issues/1297

**Comment by ArthurHDN** [comment]
Hi.
Reading https://github.com/bitcraze/crazyflie-firmware/issues/1297 solved this behavior. All good now. Thanks!
This is how it looks like:
```python
        print('Flying...')
        try:
            scf.cf.param.set_value('stabilizer.estimator', 3)
            scf.cf.param.set_value('stabilizer.controller', 5) # OOT controller
            for i in range(600):
                scf.cf.commander.send_hover_setpoint(0, 0, 0, 0) # Send dummy setpoints to my controller since the reference is hardcoded for now
                time.sleep(0.1)
        finally:
            scf.cf.commander.send_stop_setpoint()
            scf.cf.commander.send_notify_setpoint_stop() # Change to high level commander
            scf.cf.param.set_value('stabilizer.controller', 1) # Change to another controller since mine has the reference hardcoded for now
        print('Done')
```
